fs/source/source.go

```diff
@@ -54,15 +54,12 @@ import (
 // about the blob.
 type GetSources func(labels map[string]string) (source []Source, err error)
 
-// RegistryHosts returns a list of registries that provides the specified image.
-type RegistryHosts func(reference.Spec) ([]docker.RegistryHost, error)
```

Comment:
I think this might have been pseudo-duplicated to make CRI credentials work. Since something upstream from SOCI manages credentials, there could be unique credentials per image rather than per-host.

Reply:
Oh right, this would actually break on multi-image pulls with registries that support OAuth2. Meaning if we were trying to pull 2 images (`repo1`,`repo2`) on the same host `host1`. we would initially get the repo-scoped token for `repo1` and store it using host key `host1`. The subsequent image pull for `repo2` would overwrite the credentials in the cred store, so attempts to fetch the SOCI index/zTOC/spans/layers for `repo1`, we would fail since we would be providing the incorrect token (the token for `repo2`). 

So yeah, we should reference credentials by image ref/repo. I'll revert this. Thanks for the call out!

Reply:
I think I may have jumped the gun here. It looks like the `docker.Authorizer` was built with the intention of fetching tokens at _request_ time, rather than at _handle_ time. I think the idea is that the underlying credential function returns the needed authentication (not authorization), and then when attempting to authorize the request, the authorizer will take those creds and fetch the access token. 

The `AuthConfig` does have a field for the actual access token, which  the CRI completely ignores. If an `IdentityToken` (aka refresh token) or creds (username and password) is provided it will use those to fetch the access token **again** when authorizing the request.

This _may_ change in the future, and the authorizer may be extended to support directly using OAuth2 tokens from a credential func, so I still do think it's a good idea for our credential providers to be able to handle fetching/indexing scoped credentials.